### PR TITLE
[ROCm] Enable DeepSeek V4 FP8/FP4 activation quantization

### PR DESCRIPTION
**Pull request by skyguan92**
## What this changes

This enables TileLang's existing `examples/deepseek_v4/act_quant.py` FP8 and
FP4 activation-quantization paths on ROCm without changing the example:

- require `apache-tvm-ffi>=0.1.12` and use its backend-aware native C-DLPack
  exchange for ROCm E4M3FN, E4M3FNUZ, and packed E2M1x2 tensors, while retaining
  TileLang's MPS-only C-DLPack disable;
- align Python FP8 dtype selection with `hip_fp8.h`: gfx940/gfx941/gfx942 use
  FNUZ, while gfx950/gfx11/gfx12 use FN, and derive the quantization bound from
  the selected dtype;
- make the existing software FP4 E2M1 helpers available to HIP targets beyond
  gfx950;
- correctly lower scalar FP4 conversion and packed vector FP4 loads/stores;
- emit exact hexadecimal HIP floating-point constants so quantization boundary
  values round identically to TIR;
- add unit coverage plus a ROCm DeepSeek V4 activation-quantization integration
  test.

## Why

On the original base (`6b81bb8735743f14e2ecc0bdfdeaa19f24661e13`), the
DeepSeek V4 example compiles its FP8 kernel on gfx1100 but cannot exchange the
narrow-float output through the configured FFI path. The generated kernel also
uses E4M3FN bytes while Python selects E4M3FNUZ. The FP4 path additionally needs
packed vector access and the existing software FP4 helpers outside the gfx950
guard.

Review identified that an adapter-side byte-view workaround was the wrong
layer. The final implementation removes that workaround, upgrades the tvm-ffi
lower bound to 0.1.12, keeps ROCm on the native C-DLPack path, and preserves the
common FFI adapter without AMD-specific argument rewriting.

HIP constants were also emitted with the default stream precision. For FP4
scaling, `1.0f / 6.0f` became `1.666667e-01f`, changing exact boundary cases
such as `amax == 3.0` from scale `0.5` to `1.0`. Hexadecimal literals preserve
the original TIR value.

## Real-device validation

Latest exact integration gate:

- PR head: `a2c3405de38baede7e6ad05080178a54fa4373c4`;
- official main tested: `14bac09d53455f30175aed039957ab3c16ee8407`;
- local and GitHub synthetic merge tree: `e721ba6cf99ab5423574640c056f3bb4febebed1`;
- machine: eight native Radeon PRO W7900D-class gfx1100 devices, no
  `HSA_OVERRIDE_GFX_VERSION` or FFI bypass variable;
- PyTorch `2.9.1+gitff65f5b`, HIP `7.2.53211-e1a6bc5663`, tvm-ffi 0.1.12.

The exact merge tree was configured with `USE_ROCM=ON`, `USE_CUDA=OFF` and
built from source (688 Ninja steps). Static FFI/language coverage passes
30/30, Ruff and clang-format pass, and the native gates report:

```text
native device / C-DLPack metadata: E4M3FN, E4M3FNUZ, E2M1x2 exact

$ python examples/deepseek_v4/act_quant.py
[PASS] test_fp8_act_quant M=256, N=1024, block_size=128
  (FP4 bit-exact match: 131072 bytes)
[PASS] test_fp4_act_quant M=256, N=1024, block_size=32
  FP8 round-trip MSE: 0.000706
  FP4 round-trip MSE: 0.013256
[PASS] test_round_trip_error

10/10 seeds: zero FP4 scale or packed-byte mismatch
31/31 passed: focused native FP8/FP4, FFI, and language suite
77/77 passed: HIP codegen and vector-Select suite
```

Since that gate, upstream main advanced five commits without touching any of
the PR's 16 paths. The public head remains mergeable and `CLEAN`; all 12 checks
succeed and two publish-only jobs are expected skips.

## Review follow-up

- `bae2102e` scopes FP4 vector-load SSA temporaries and added the then-relevant
  non-contiguous fallback boundary test.
- `9be6a1a2` absorbs the maintainer feedback: removes AMD-specific argument
  rewriting from the common FFI adapter, adopts tvm-ffi 0.1.12, keeps ROCm on
  native C-DLPack, retains only the MPS disable, derives FN/FNUZ bounds from the
  selected dtype, and documents the packed vector-store branch.
- later commits make FP4 header inclusion explicit for packed stores, preserve
  exact float constants, and extend the native regression matrix.

All current review threads are resolved. The latest real-device gate above is
on the exact current merge tree, not merely the original review commit.

## Scope

This adds a software FP4 activation-quantization/cast/copy foundation. It does
**not** claim or add gfx1100 hardware FP4 matrix instructions, an FP4 weight
GEMM, checkpoint-quality validation, or end-to-end model throughput.

## Summary

- Enables ROCm FP8 and software FP4 activation quantization for DeepSeek V4.
- Uses tvm-ffi 0.1.12 native C-DLPack exchange for ROCm narrow-float tensors.
- Updates ROCm FP8 dtype selection across AMD GPU targets.
- Adds portable FP4 E2M1 helpers and HIP lowering for scalar and packed vector conversions.
- Adds exact hexadecimal floating-point constant emission.
- Adds ROCm unit tests, DLPack metadata tests, environment tests, and a DeepSeek V4 integration test.

This PR does not add FP4 matrix instructions or an FP4 GEMM path.

## C++ style / lint notes

- The PR changes C++ code in `src/rocm/codegen`.
- It does not change `docs/developer_guide/cpp_style.md`.
- The **C++ API Style Audit (warning only)** runs in CI.
- TLCPP003/TLCPP004 findings are advisory. They should not block merge unless they identify a correctness, FFI, or maintainability risk.

### CODE REVIEW

File: src/rocm/codegen/codegen_hip.cc [hunk 1]
```diff
@@ -1830,6 +1846,63 @@ void CodeGenTileLangHIP::VisitStmt_(const AllocBufferNode *op) {
...
+  stream << ' ' << result << "{};\n";
+  for (int i = 0; i < value_dtype.lanes(); ++i) {
+    std::ostringstream lane_index;
+    print_lane_index(i, lane_index);
+    PrintVecElemStore(result, value_dtype, i,
+                      "tl_fp4_packed_load(" + packed_buffer + ", " +
+                          lane_index.str() + ")");
+  }
+  os << result;
+}
```
[the remark this answers is not in the document]
Comment (skyguan92) [reply src/rocm/codegen/codegen_hip.cc:2003]: Fixed in bae2102e. The FP4 vector-load path now wraps its SSAGetID temporaries with BeginScope/EndScope, matching the store path and PrintVecBinaryOp. On a physical W7900/gfx1100, the incremental HIP build passes, the unmodified DeepSeek V4 activation-quantization example remains FP4 bit-exact for all 131072 bytes, 10 fixed seeds have zero scale/byte mismatches, the focused suite is 13/13, and the existing HIP-codegen suite is 36/36.

File: tilelang/jit/adapter/tvm_ffi.py [hunk 2]
```diff
@@ -27,6 +27,37 @@
...
+        )
+
+    # The Python DLPack fallback supports byte tensors, while FP8/FP4 are not
+    # portable DLPack dtypes. Reinterpret the same storage on both sides while
+    # retaining the Torch storage dtype and physical shape. In particular, the
+    # generated host wrapper expects packed FP4x2 as a subtype of scalar FP4;
+    # it validates the equivalent total bit count and physical strides.
+    byte_tensor = tensor.view(torch.int8)
+    tvm_tensor = runtime.from_dlpack(torch.utils.dlpack.to_dlpack(byte_tensor))
+    return tvm_tensor._create_view(list(tensor.shape), dtype=str(storage_dtype))
```
[the remark this answers is not in the document]
Comment (skyguan92) [reply tilelang/jit/adapter/tvm_ffi.py]: Fixed in bae2102e. The narrow-float DLPack bridge now rejects non-contiguous tensors with a clear ValueError before the dtype-changing byte view, and a transposed-float8 unit test covers that boundary. The contiguous path remains zero-copy and the existing round-trip test still verifies the original data_ptr. Physical W7900 validation passes: 13/13 focused tests, the unmodified DeepSeek V4 example, and 36/36 existing HIP-codegen tests.

File: tilelang/jit/adapter/tvm_ffi.py [hunk 3]
```diff
@@ -274,7 +312,14 @@ def func(*inputs: torch.Tensor | Any):
                 tensor_list.append(tensor)
 
             executable = self._get_executable()
-            executable(*tensor_list)
+            if rocm_narrow_float_mask is None:
+                executable(*tensor_list)
+            else:
+                runtime_args = [
```
Comment (LeiWang1999) [inline tilelang/jit/adapter/tvm_ffi.py]: Thanks, but these codes are not very elegant in my opinion, I'll take a look then

Comment (LeiWang1999) [reply tilelang/jit/adapter/tvm_ffi.py]: It looks like this extra code exists to work around `torch.to_dlpack` not being
able to convert FP8/FP4 dtypes on ROCm. However, tvm-ffi itself already has
this conversion — the error only occurs because of two things:

1. tvm-ffi's `prefer_rocm_override` logic, introduced by
   [apache/tvm-ffi#466](https://github.com/apache/tvm-ffi/pull/466) to force its
   own DLPack addon over the upstream PyTorch exchange API on ROCm (a stream-
   correctness fix that came with an unrelated side effect), and already
   removed in
   [apache/tvm-ffi#585](https://github.com/apache/tvm-ffi/pull/585)
   ("Prefer upstream PyTorch DLPack API in torch extension loader"). The
   override is gone starting from
   [tvm-ffi v0.1.12](https://github.com/apache/tvm-ffi/releases/tag/v0.1.12).

2. TileLang itself disabling the ROCm C DLPack path in
   [tile-ai/tilelang#2179](https://github.com/tile-ai/tilelang/pull/2179)
   (`_disable_tvm_ffi_torch_c_dlpack` in `tilelang/__init__.py`), which was a
   workaround for tensor-metadata corruption seen through the C DLPack fast
   path on ROCm ("ndim expected 2, but got 512").

Since the underlying issue should be resolved in the latest tvm-ffi
(>= v0.1.12), we should upgrade tvm-ffi, re-enable the ROCm C DLPack path in
TileLang, and remove this host-side byte-view bridge instead of patching
around it here.

File: src/rocm/codegen/codegen_hip.cc [hunk 4]
```diff
@@ -1839,6 +1914,49 @@ void CodeGenTileLangHIP::VisitStmt_(const BufferStoreNode *op) {
   DataType element_dtype = op->buffer->dtype;
   Var buffer_var = op->buffer->data;
 
+  if (element_dtype.is_float4() && element_dtype.is_scalar() &&
```
Comment (SiriusNEO) [inline src/rocm/codegen/codegen_hip.cc:2016]: Could you add a comment here to brief describe this branch?

File: tilelang/jit/adapter/tvm_ffi.py [hunk 5]
```diff
@@ -27,6 +27,43 @@
 
 COMPILE_ARGS = {}
 
+
+def _rocm_narrow_float_param_mask(params: list[KernelParam]) -> tuple[bool, ...] | None:
```
Comment (SiriusNEO) [inline tilelang/jit/adapter/tvm_ffi.py]: We don't encourage adding AMD-related code in the common namespace (ffi side). Are they necessary to be put in this place?

**Review by SiriusNEO — COMMENTED** [review]
Left some comments. And one concern: `testing/python/amd/test_tilelang_deepseek_v4_act_quant.py` uses `fp8_max = 448.0`, which only works for `fn` (for `fnuz`, it's `240` I think?)

File: tilelang/jit/adapter/tvm_ffi.py (hunk 3 above)
Comment (skyguan92) [reply tilelang/jit/adapter/tvm_ffi.py]: Addressed in `9be6a1a2`. I removed `_rocm_narrow_float_param_mask`, `_export_rocm_narrow_float_as_tvm_view`, and the adapter-side runtime argument rewrite; ROCm now keeps the C DLPack path enabled and `TVMFFIKernelAdapter` passes tensors directly. All declared dependency and install lower bounds are now `apache-tvm-ffi>=0.1.12,<0.1.13`; the existing MPS-only disable path is retained and regression-tested.

With both disable variables explicitly unset on a physical W7900/gfx1100, tvm-ffi 0.1.12 preserved the exact C-FFI shape/dtype/device metadata for E4M3FN, E4M3FNUZ, and packed E2M1x2. The unmodified DeepSeek V4 example passed FP8; FP4 matched all 131072 packed bytes; 10/10 deterministic seeds had zero scale/byte mismatches; the focused suite was 28/28; and the existing HIP-codegen suite was 36/36.

File: src/rocm/codegen/codegen_hip.cc (hunk 4 above)
Comment (skyguan92) [reply src/rocm/codegen/codegen_hip.cc:2016]: Addressed in `9be6a1a2`. The branch comment now explains that vector FP4 values targeting scalar FP4 buffers require lane-wise packed stores because CodeGenC generic scalarization advances one byte per element. `clang-format --dry-run --Werror` passes, and the exact W7900 candidate passes the unmodified FP4 example plus the existing HIP-codegen suite 36/36.

File: tilelang/jit/adapter/tvm_ffi.py (hunk 5 above)
Comment (skyguan92) [reply tilelang/jit/adapter/tvm_ffi.py]: Addressed in `9be6a1a2`. The AMD-specific mask, byte-view exporter, and runtime argument rewrite are removed from the common FFI adapter. ROCm now relies on the backend-aware tvm-ffi 0.1.12 exchange path; only the existing MPS-specific disable remains in `tilelang/__init__.py`. The physical-W7900 native-path gate passed the full example, 10 exact seeds, 28/28 focused tests, and 36/36 HIP-codegen tests.

**Comment by skyguan92** [comment]
Follow-up on the review-body `fp8_max` concern: addressed in `9be6a1a2`. Both the generated kernel clamp and the Python reference now derive the finite maximum from the target-selected `T.float8_e4m3.as_torch()` dtype via `torch.finfo`, rather than hard-coding 448. The parametrized target-selection regression requires gfx942/FNUZ = 240 and gfx950, gfx1100, and gfx1201/FN = 448.

On the physical W7900/gfx1100 path, the selected FN maximum is 448 and the unmodified example remains exact: FP8 passes, FP4 is 131072/131072 bytes exact, and round-trip MSE remains 0.000706/0.013256. This W7900 run does not claim a physical gfx942 result; the FNUZ limit is covered by target-selection and native C-FFI metadata tests.

**Comment by skyguan92** [comment]
Addressed the latest CodeRabbit outside-diff docstring finding in `b2e219b8`: the reference return description now says it uses the target-selected E4M3 FN/FNUZ variant instead of hard-coding `torch.float8_e4m3fn`. This is a one-line documentation-only change; the exact new head still passes the full fail-closed W7900 gate (27/27 CPU, FP4 131072/131072 bytes exact, 10/10 seeds exact, 28/28 focused, 36/36 HIP-codegen).

**Review by SiriusNEO — DISMISSED** [review]
Overall LGTM, but it seems like this PR requires `ffi>=0.1.12` since it needs some AMD-aware changes in this version. cc @LeiWang1999

**Comment by skyguan92** [comment]
Thanks — agreed. The current head already raises the runtime floor to `apache-tvm-ffi>=0.1.12,<0.1.13` in `pyproject.toml`, `requirements.txt`, `requirements-dev.txt`, and the manual ROCm installation command (commit `9be6a1a2`). The W7900 validation also used tvm-ffi 0.1.12 through its native C DLPack exchange path. No additional source change is needed for this requirement.

**Comment by skyguan92** [comment]
Current-main compatibility and the latest review finding are addressed in `bb0b9809` and `7c1a1a7f`.

- The branch now three-way merges cleanly with current `main` (`12dbf3e9`); GitHub reports it mergeable. The final merge result changes the intended 17 files only.
- CodeRabbit's FP4-header finding was valid: the vector packed-store path now sets `enable_fp4_` before emitting `tl_fp4_packed_store`.
- HIP constants keep the legacy decimal form only when it parses back to the exact TIR value with the same sign. This preserves current-main's unmodified decimal zero/one Select assertions while values such as `1/6` remain exact hexadecimal literals, so the quantization-boundary fix is retained.

Physical W7900D/gfx1100 validation on the exact final merge source (Torch 2.9.1, HIP 7.2, tvm-ffi 0.1.12, native C-DLPack, no architecture override):

- ROCm-only incremental rebuild passed (`USE_ROCM=ON`, `USE_CUDA=OFF`).
- Current-main HIP codegen + vector-Select suite: 77/77 passed.
- Native focused suite: 28/28 passed.
- Native metadata preserved E4M3FN, E4M3FNUZ, and packed E2M1x2.
- In-tree DeepSeek V4 example: FP4 131072/131072 bytes exact; FP8/FP4 round-trip MSE 0.000706/0.013256.
- Ten fixed FP4 seeds: zero scale mismatches and zero byte mismatches.
- Generated source includes `hip_fp4.h` and emits `1/6` as exact hex.
- Two pre-run and one final idle samples: all 8 GPUs at 0% use, 28,004,352 bytes VRAM each, no KFD PID.

`pre-commit.ci` and CodeRabbit pass on `7c1a1a7f`; repository CI/Dist are awaiting workflow approval. Ready for maintainer re-review when convenient.
